File: js/src/core/indexer.ts
/**
 * hanimo-rag v2 — LLM-based Key Extraction Indexer
 *
 * Reads each text chunk, sends it to an LLM, and extracts structured
 * key-value tags (topics, entities, questions, category, summary).
 */

import * as crypto from 'node:crypto'
import type { LLMBase } from '../llm/base.js'
import type { StoreBase } from '../store/base.js'
import type { ExtractedKeys, IndexedChunk, IndexResult, ParsedDocument } from '../types.js'
import { chunkText } from './chunker.js'

const EXTRACTION_SYSTEM = `You are a precise document indexer. Given a text chunk, extract structured metadata for search indexing.

Return a JSON object with these exact fields:
- "topics": array of 3-8 key topics/themes (lowercase, short phrases)
- "entities": array of named entities — people, organizations, technologies, products, places (as mentioned)
- "questions": array of 2-5 questions this chunk could answer
- "category": one of: "technical", "business", "tutorial", "reference", "narrative", "data", "other"
- "summary": one-sentence summary of the chunk (max 100 words)`

function buildExtractionPrompt(content: string): string {
  return `Extract indexing metadata from this text chunk:

---
${content}
---

Respond with a JSON object containing: topics, entities, questions, category, summary.`
}

export class Indexer {
  private llm: LLMBase
  private store: StoreBase

  constructor(llm: LLMBase, store: StoreBase) {
    this.llm = llm
    this.store = store
  }

  async indexDocument(
    source: string,
    doc: ParsedDocument,
    options: { chunkSize?: number; chunkOverlap?: number } = {},
  ): Promise<{ chunks: number }> {
    const chunkSize = options.chunkSize ?? 1000
    const chunkOverlap = options.chunkOverlap ?? 200

    const textChunks = chunkText(doc.text, { chunkSize, chunkOverlap })

    if (textChunks.length === 0) {
      return { chunks: 0 }
    }

    const indexedChunks: IndexedChunk[] = []

    for (const content of textChunks) {
      // Skip very small chunks
      if (content.trim().length < 20) {
        continue
      }

      const keys = await this.extractKeys(content)

      const chunk: IndexedChunk = {
        id: crypto.randomUUID(),
        source,
        content,
        topics: keys.topics,
        entities: keys.entities,
        questions: keys.questions,
        category: keys.category,
        summary: keys.summary,
        indexedAt: new Date().toISOString(),
      }

      indexedChunks.push(chunk)
    }

    if (indexedChunks.length > 0) {
      this.store.saveChunks(indexedChunks)
    }

    return { chunks: indexedChunks.length }
  }

  async indexMultiple(
    documents: Array<{ source: string; doc: ParsedDocument }>,
    options: { chunkSize?: number; chunkOverlap?: number } = {},
  ): Promise<IndexResult> {
    const result: IndexResult = { indexed: 0, files: [] }

    for (const { source, doc } of documents) {
      const { chunks } = await this.indexDocument(source, doc, options)
      result.indexed += chunks
      result.files.push({ file: source, chunks })
    }

    return result
  }

  private async extractKeys(content: string): Promise<ExtractedKeys> {
    const prompt = buildExtractionPrompt(content)

    try {
      const extracted = await this.llm.generateJson<ExtractedKeys>(prompt, EXTRACTION_SYSTEM)
      return normalizeKeys(extracted)
    } catch {
      // Fallback: extract basic keys from content without LLM
      return fallbackExtraction(content)
    }
  }
}

function normalizeKeys(raw: Partial<ExtractedKeys>): ExtractedKeys {
  return {
    topics: ensureStringArray(raw.topics).slice(0, 10),
    entities: ensureStringArray(raw.entities).slice(0, 10),
    questions: ensureStringArray(raw.questions).slice(0, 5),
    category: typeof raw.category === 'string' ? raw.category : 'other',
    summary: typeof raw.summary === 'string' ? raw.summary : '',
  }
}

function ensureStringArray(val: unknown): string[] {
  if (!Array.isArray(val)) return []
  return val.filter((item): item is string => typeof item === 'string')
}

/**
 * Fallback key extraction when LLM fails.
 * Uses simple heuristics to extract basic metadata.
 */
function fallbackExtraction(content: string): ExtractedKeys {
  // Extract "important" words (longer words that might be topics)
  const words = content
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, ' ')
    .split(/\s+/)
    .filter((w) => w.length > 4)

  // Count word frequency
  const freq = new Map<string, number>()
  for (const w of words) {
    freq.set(w, (freq.get(w) ?? 0) + 1)
  }

  const topics = [...freq.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5)
    .map(([word]) => word)

  // Extract capitalized words as potential entities
  const entityPattern = /\b[A-Z][a-zA-Z]+(?:\s+[A-Z][a-zA-Z]+)*\b/g
  const entityMatches = content.match(entityPattern) ?? []
  const entities = [...new Set(entityMatches)].slice(0, 5)

  const firstSentence = content.split(/[.!?]/)[0]?.trim() ?? ''
  const summary = firstSentence.length > 100 ? firstSentence.slice(0, 97) + '...' : firstSentence

  return {
    topics,
    entities,
    questions: [],
    category: 'other',
    summary,
  }
}

File: hanimo_rag/core/indexer.py
"""LLM-based document indexer — the core innovation of HanimoRAG."""
from __future__ import annotations

import logging
from pathlib import Path

from hanimo_rag.core.chunker import chunk_text
from hanimo_rag.llm.base import LLMBase
from hanimo_rag.parsers import parse_file
from hanimo_rag.store.base import StoreBase
from hanimo_rag.types import IndexedChunk

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = (
    "You are a document indexer. Extract structured metadata from text chunks "
    "for search indexing. Always output valid JSON."
)

EXTRACT_PROMPT = """Analyze this text and extract metadata for search indexing.

<text>
{text}
</text>

Output JSON only:
{{
  "topics": ["keyword1", "keyword2", ...],
  "entities": ["Entity1", "Entity2", ...],
  "questions": ["question this text answers", ...],
  "category": "tutorial|reference|api|concept|troubleshooting|config|example|discussion|news|other",
  "summary": "one sentence summary"
}}"""


class Indexer:
    """Index documents by extracting LLM-generated metadata for each chunk."""

    def __init__(self, llm: LLMBase, store: StoreBase) -> None:
        self.llm = llm
        self.store = store

    async def index_file(
        self,
        file_path: str,
        chunk_size: int = 512,
        chunk_overlap: int = 50,
    ) -> dict:
        """Parse and index a file."""
        path = Path(file_path)
        if not path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")

        doc = parse_file(file_path)
        if not doc.text.strip():
            logger.warning("Empty document: %s", file_path)
            return {"source": file_path, "chunks": 0, "status": "empty"}

        return await self.index_text(
            doc.text,
            source=file_path,
            chunk_size=chunk_size,
            chunk_overlap=chunk_overlap,
        )

    async def index_text(
        self,
        text: str,
        source: str = "manual",
        chunk_size: int = 512,
        chunk_overlap: int = 50,
    ) -> dict:
        """Index raw text by chunking and extracting metadata."""
        text_chunks = chunk_text(text, chunk_size=chunk_size, chunk_overlap=chunk_overlap)

        if not text_chunks:
            return {"source": source, "chunks": 0, "status": "empty"}

        indexed_chunks: list[IndexedChunk] = []
        errors = 0

        for tc in text_chunks:
            try:
                metadata = await self._extract_metadata(tc.text)
                chunk_id = IndexedChunk.make_id(source, tc.text)

                indexed_chunks.append(
                    IndexedChunk(
                        id=chunk_id,
                        source=source,
                        content=tc.text,
                        topics=metadata.get("topics", []),
                        entities=metadata.get("entities", []),
                        questions=metadata.get("questions", []),
                        category=metadata.get("category", "other"),
                        summary=metadata.get("summary", ""),
                    )
                )
            except Exception as exc:
                logger.error("Failed to extract metadata for chunk %d: %s", tc.index, exc)
                errors += 1
                # Still save the chunk with minimal metadata
                chunk_id = IndexedChunk.make_id(source, tc.text)
                indexed_chunks.append(
                    IndexedChunk(
                        id=chunk_id,
                        source=source,
                        content=tc.text,
                    )
                )

        saved = self.store.save_chunks(indexed_chunks)

        return {
            "source": source,
            "chunks": len(indexed_chunks),
            "saved": saved,
            "errors": errors,
            "status": "ok",
        }

    async def _extract_metadata(self, text: str) -> dict:
        """Use LLM to extract structured metadata from a text chunk."""
        prompt = EXTRACT_PROMPT.format(text=text[:2000])  # Truncate very long chunks
        result = await self.llm.generate_json(prompt, system=SYSTEM_PROMPT)

        # Validate and normalize
        return {
            "topics": _ensure_str_list(result.get("topics", [])),
            "entities": _ensure_str_list(result.get("entities", [])),
            "questions": _ensure_str_list(result.get("questions", [])),
            "category": str(result.get("category", "other")),
            "summary": str(result.get("summary", "")),
        }


def _ensure_str_list(value: object) -> list[str]:
    """Ensure a value is a list of strings."""
    if isinstance(value, list):
        return [str(v) for v in value if v]
    if isinstance(value, str):
        return [value]
    return []
